flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/statistics/basicstatistic/MultivariateGaussian.java

```diff
@@ -0,0 +1,148 @@
+			eps /= 2.0;
+		}
+		EPSILON = eps;
+	}
+
+	private final DenseVector mean;
+	private final DenseMatrix cov;
+
+	private DenseMatrix rootSigmaInv;
+	private double u;
+
+	// data buffers for computing pdf
+	private DenseVector delta;
+	private DenseVector v;
+
+	/**
+	 * The constructor.
+	 *
+	 * @param mean The mean vector of the distribution.
+	 * @param cov  The covariance matrix of the distribution.
+	 */
+	public MultivariateGaussian(DenseVector mean, DenseMatrix cov) {
+		this.mean = mean;
+		this.cov = cov;
+		this.delta = DenseVector.zeros(mean.size());
+		this.v = DenseVector.zeros(mean.size());
+		calculateCovarianceConstants();
+	}
+
+	/**
+	 * Returns density of this multivariate Gaussian at given point, x.
+	 */
+	public double pdf(Vector x) {
+		return Math.exp(logpdf(x));
+	}
+
+	/**
+	 * Returns the log-density of this multivariate Gaussian at given point, x.
+	 */
+	public double logpdf(Vector x) {
+		int n = mean.size();
+		System.arraycopy(mean.getData(), 0, delta.getData(), 0, n);
+		BLAS.scal(-1.0, delta);
+		if (x instanceof DenseVector) {
+			BLAS.axpy(1., (DenseVector) x, delta);
+		} else if (x instanceof SparseVector) {
+			BLAS.axpy(1., (SparseVector) x, delta);
+		}
+		BLAS.gemv(1.0, rootSigmaInv, false, delta, 0., v);
+		return u - 0.5 * BLAS.dot(v, v);
+	}
+
+	/**
+	 * Compute distribution dependent constants.
+	 *
+	 * <p>The probability density function is calculated as:
+	 * pdf(x) = (2*pi)^(-k/2)^ * det(sigma)^(-1/2)^ * exp((-1/2) * (x-mu).t * inv(sigma) * (x-mu))
+	 *
+	 * <p>Here we compute the following distribution dependent constants that can be reused in each pdf computation:
+	 * A) u = log((2*pi)^(-k/2)^ * det(sigma)^(-1/2)^)
+	 * B) rootSigmaInv = sqrt(inv(sigma)) = D^(-1/2)^ * U.t
+	 *
+	 * <p><ul>
+	 * <li> sigma = U * D * U.t
+	 * <li> inv(sigma) = U * inv(D) * U.t = (D^{-1/2}^ * U.t).t * (D^{-1/2}^ * U.t)
+	 * <li> sqrt(inv(sigma)) = D^(-1/2)^ * U.t
+	 * </ul></p>
+	 */
+	private void calculateCovarianceConstants() {
+		int n = this.mean.size();
+		int lwork = 3 * n - 1;
+		double[] matA = new double[n * n];
+		double[] work = new double[lwork];
+		double[] evs = new double[n];
+		intW info = new intW(0);
+
+		for (int i = 0; i < n; i++) {
+			System.arraycopy(cov.getData(), i * n, matA, i * n, i + 1);
+		}
+		LAPACK_INST.dsyev("V", "U", n, matA, n, evs, work, lwork, info);
+
+		double maxEv = -Double.MAX_VALUE;
+		for (int i = 0; i < n; i++) {
+			if (evs[i] > maxEv) {
+				maxEv = evs[i];
+			}
+		}
+		double tol = EPSILON * n * maxEv;
+
+		// log(pseudo-determinant) is sum of the logs of all non-zero singular values
+		double logPseudoDetSigma = 0.;
+		for (double ev : evs) {
+			if (ev > tol) {
+				logPseudoDetSigma += Math.log(ev);
+			}
+		}
+
+		this.rootSigmaInv = new DenseMatrix(n, n);
+		for (int i = 0; i < n; i++) {
+			double pinvS = evs[i] > tol ? Math.sqrt(1.0 / evs[i]) : 0.;
+			for (int j = 0; j < n; j++) {
+				this.rootSigmaInv.set(i, j, pinvS * matA[j * n + i]);
```

Comment:
is there a breeze routine that can speed this up. this is basically a scale up from a pseudo inverse of one element of the eigenvector rigth?

Reply:
Yes, you are right. Here we are just scaling the eigen vectors. Now I use BLAS's scal routine to do this for me. More over, by switching rootSigmaInv from D^(-1/2)^ * U.t to U * D^(-1/2)^, we can reuse the data buffer of U.